lib/handleRequest.js

```diff
@@ -125,29 +128,42 @@ function validationCompleted (request, reply, validationErr) {
 function preHandlerCallback (err, request, reply) {
   if (reply.sent) return
 
+  const store = { request, reply }
+
   if (err != null) {
     reply[kReplyIsError] = true
     reply.send(err)
+    store.error = err
+    channels.end.publish(store)
     return
   }
 
   let result
 
+  channels.start.publish(store)
+
   try {
     result = request[kRouteContext].handler(request, reply)
   } catch (err) {
+    store.error = err
+    channels.error.publish(store)
+
     reply[kReplyIsError] = true
     reply.send(err)
+    channels.end.publish(store)
```

Comment:
Rather than putting the `end` in all the branches you could use try/finally. That also ensures the event still fires if something throws.